### PR TITLE
Improve root message handling and interface

### PR DESCRIPTION
**Pull request by jakekidd**
- Added immutable flag `IS_HUB` set on instantiation.
- Renamed `sendMessage` / `receiveMessage` to `sendRoot` / `receiveRoot` to make interface less confusing (we already have `process` methods)
- Renamed related events as well.
- `sendRoot` takes no arguments and just sends the current `outboundRoot`
- Virtual method `_receiveRoot`, which should be implemented by the specific <Domain>Connector, now returns the new updated root (`bytes32`). `receiveRoot` calls internal `_receiveRoot`, sets the `aggregateRoot` to the returned value, and if `IS_HUB` is true, will also update the RootManager (by calling `setOutboundRoot(mirrorDomain, aggregateRoot)`)
- Added method `propagateRoot` which can only be called by `RootManager` on the hub chain (i.e. if `IS_HUB` is true)

### CODE REVIEW

File: packages/deployments/contracts/contracts/core/messaging/interfaces/IConnector.sol [hunk 1]
```diff
@@ -40,11 +40,23 @@ import {IProposedOwnable} from "../../shared/interfaces/IProposedOwnable.sol";
...
+   * chain out to the spoke/L2 chains.
+   */
+  function propagateRoot(bytes memory _data) external;
 
-  function processMessage(address _sender, bytes memory _data) external;
+  /**
+   * @notice This is called by relayers to trigger passing of current root to mainnet root manager
+   * @dev This is called at specific time intervals
+   */
+  function sendRoot() external;
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/interfaces/IConnector.sol:53]: I don't love having two different functions that are designed to both send a `bytes32` to their mirror. it seems like we could make it a single function and use the `IS_HUB` tag

File: packages/deployments/contracts/contracts/core/messaging/connectors/MainnetConnector.sol [hunk 2]
```diff
@@ -32,7 +43,7 @@ contract MainnetL1Connector is Connector {
    * @dev Messaging uses this function to send outboundRoot to root manager. On mainnet,
    * would just update the root manager directly
    */
-  function _sendMessage(bytes memory _data) internal override {
+  function _sendRoot(bytes memory _data) internal override {
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/MainnetConnector.sol:46]: needs `msg.sender` checks as well

File: packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol [hunk 3]
```diff
@@ -222,21 +236,51 @@ abstract contract Connector is ProposedOwnable, MerkleTreeManager, ConnectorMana
...
-  function sendMessage(bytes memory _data) external {
-    _sendMessage(_data);
+  /**
+   * @notice This should be called by RootManager to propagate an updated aggregate root on the Hub/L1
+   * chain out to the spoke/L2 chains.
+   */
+  function propagateRoot(bytes memory _data) external onlyRootManager onlyHub {
+    // RootManager has aggregated roots from all other Connectors on this hub chain.
+    // TODO: Should we update our current aggregateRoot to reflect the new one.
+    // aggregateRoot = bytes32(_data);
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol:246]: yeah -- i don't see why not

File: packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol [hunk 4]
```diff
@@ -40,9 +41,9 @@ abstract contract Connector is ProposedOwnable, MerkleTreeManager, ConnectorMana
 
   event Process(bytes32 leaf, bool success, bytes returnData);
 
-  event MessageSent(bytes data, address caller);
+  event RootUpdateSent(bytes data, address caller);
 
-  event MessageProcessed(address from, bytes data, address caller);
+  event RootUpdateReceived(address from, bytes data, address caller);
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol:46]: do we want to be more specific with outbound / aggregate roots in the events if we have functions that are more opinionated?

File: packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol [hunk 5]
```diff
@@ -222,21 +236,51 @@ abstract contract Connector is ProposedOwnable, MerkleTreeManager, ConnectorMana
...
+  /**
+   * @notice This should be called by RootManager to propagate an updated aggregate root on the Hub/L1
+   * chain out to the spoke/L2 chains.
+   */
+  function propagateRoot(bytes memory _data) external onlyRootManager onlyHub {
+    // RootManager has aggregated roots from all other Connectors on this hub chain.
+    // TODO: Should we update our current aggregateRoot to reflect the new one.
+    // aggregateRoot = bytes32(_data);
+
+    // TODO: This code block is partially redundant with the `sendRoot` method. Should we just make `sendRoot` public?
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol:248]: `sendRoot` is public? im confuse

File: packages/deployments/contracts/contracts/core/messaging/interfaces/IConnector.sol (hunk 1 above)
Comment (LayneHaber) [reply packages/deployments/contracts/contracts/core/messaging/interfaces/IConnector.sol:53]: is the issue the difference in args? if so, then lets be more explicit in the name about the roots we are sending. `sendOutboundRoot` / `sendAggregateRoot` or something

Comment (jakekidd) [reply packages/deployments/contracts/contracts/core/messaging/interfaces/IConnector.sol:53]: down for that! the second thing, that is. `sendOutboundRoot` and `sendAggregateRoot`

File: packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol (hunk 5 above)
Comment (jakekidd) [reply packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol:248]: it's `external` and `external != public` lol.

I was thinking - should we just make sendRoot public, then we can technically call it externally *and* internally but meh

File: packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol (hunk 4 above)
Comment (jakekidd) [reply packages/deployments/contracts/contracts/core/messaging/connectors/Connector.sol:46]: as in, with the naming here?

Like, `OutboundRootSent` or `AggregateRootReceived` ?

**Comment by jakekidd** [comment]
Closing this out as we accomplished with similar issues - may come back later to reuse some things from this

**Pull request closed without merging**